conformance/tests/overloads_evaluation.py

```diff
@@ -0,0 +1,224 @@
+
+@overload
+def expand_enum(x: Literal[Color.RED]) -> Literal[0]:
+    ...
+
+@overload
+def expand_enum(x: Literal[Color.BLUE]) -> Literal[1]:
+    ...
+
+def expand_enum(x: Color) -> int:
+    return x.value
+
+def check_expand_enum(v: Color) -> None:
+    ret1 = expand_enum(v)
+    assert_type(ret1, Literal[0, 1])
+
+
+# > 4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
+
+@overload
+def expand_type_union(x: type[int]) -> int:
+    ...
+
+@overload
+def expand_type_union(x: type[str]) -> str:
+    ...
+
+def expand_type_union(x: type[int] | type[str]) -> int | str:
+    return 1
+
+def check_expand_type_union(v: type[int | str]) -> None:
+    ret1 = expand_type_union(v)
+    assert_type(ret1, int | str)
+
+
+# > 5. Tuples of known length that contain expandable types should be expanded
+# > into all possible combinations of their element types. For example, the type
+# > ``tuple[int | str, bool]`` should be expanded into ``(int, Literal[True])``,
+# > ``(int, Literal[False])``, ``(str, Literal[True])``, and
+# > ``(str, Literal[False])``.
+
+@overload
+def expand_tuple(x: tuple[int, int]) -> int:
+    ...
+
+@overload
+def expand_tuple(x: tuple[int, str]) -> str:
+    ...
+
+def expand_tuple(x: tuple[int, int | str]) -> int | str:
+    return 1
+
+def check_expand_tuple(v: int | str) -> None:
+    ret1 = expand_tuple((1, v))
+    assert_type(ret1, int | str)
+
+
+# > Step 4: If the argument list is compatible with two or more overloads,
+# > determine whether one or more of the overloads has a variadic parameter
+# > (either ``*args`` or ``**kwargs``) that maps to a corresponding argument
+# > that supplies an indeterminate number of positional or keyword arguments.
+# > If so, eliminate overloads that do not have a variadic parameter.
+
+@overload
+def variadic(x: int, /) -> str:
+    ...
+
+@overload
+def variadic(x: int, y: int, /, *args: int) -> int:
+    ...
+
+def variadic(*args: int) -> int | str:
+    return 1
+
+# > - If this results in only one remaining candidate overload, it is
+# >   the winning match. Evaluate it as if it were a non-overloaded function
+# >   call and stop.
+
+def check_variadic(v: list[int]) -> None:
+    ret1 = variadic(*v)
+    assert_type(ret1, int)
+
+
+# > Step 5: For each argument, determine whether all possible
+# > :term:`materializations <materialize>` of the argument's type are assignable to
+# > the corresponding parameter type for each of the remaining overloads. If so,
+# > eliminate all of the subsequent remaining overloads.
+
+@overload
+def example4(x: list[int], y: int) -> int:
+   ...
+
+@overload
+def example4(x: list[str], y: str) -> int:
+    ...
+
+@overload
+def example4(x: int, y: int) -> list[int]:
+    ...
+
+def example4(x: list[int] | list[str] | int, y: int | str) -> int | list[int]:
+    return 1
+
+def check_example4(v1: list[Any], v2: Any):
+    ret1 = example4(v1, v2)
+    assert_type(ret1, int)
+
+    ret2 = example4(v2, 1)
+    assert_type(ret2, Any)
```

Comment:
Would it make sense to add more test cases for ambiguous overload resolution resulting in an `Any` return type? I couldn't see other tests for this.

Reply:
Are there particular distinct scenarios that you think should also be tested?

Reply:
You might want to look at the unit tests in pyright and mypy. I wouldn't recommend copying test code verbatim, but they can serve as a source of inspiration for additional conformance tests.

Here are some relevant test cases in the pyright test suite:
* Cases involving [argument type expansion](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/tests/samples/overloadCall4.py)
* Cases involving the handling of ambiguous overloads because argument [types include `Any`](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/tests/samples/overloadCall6.py)
* Cases involving [variadic parameters in overloads](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/tests/samples/overloadCall5.py)

Reply:
I don't expect to contribute more test cases to this PR myself soon, but I have no objection to anyone building on the existing test cases who sees gaps they would like to fill.

Reply:
I plan to add more test cases here, but I'd like to do that as a separate step once this PR is officially merged.